configfs/fakertmr/fakertmr.go

```diff
@@ -37,100 +38,107 @@ const (
-	rtmrMaps map[int]*rtmrValue
-}
-
 // RtmrSubsystem represents a fake configfs-tsm rtmr subsystem.
 type RtmrSubsystem struct {
 	// WriteAttr called on any WriteFile to an attribute.
-	WriteAttr func(e *rtmrEntry, attr string, contents []byte) error
+	WriteAttr func(dirname string, attr string, contents []byte) error
 	// ReadAttr is called on any non-InAddr key.
-	ReadAttr func(e *rtmrEntry, attr string) ([]byte, error)
+	ReadAttr func(dirname string, attr string) ([]byte, error)
 	// Random is the source of randomness to use for MkdirTemp
 	Random io.Reader
-	// RtmrMaps is a map of rtmr index to rtmr value.
-	RtmrMaps map[int]*rtmrValue
-	// Entries is a map of rtmr entry name to rtmr entry.
-	Entries map[string]*rtmrEntry
+	// We use a temp folder to store the rtmr entries.
+	// The path to the fake rtmr subsystem.
+	Path string
 }
 
 // RemoveAll implements configfsi.Client.
 func (r *RtmrSubsystem) RemoveAll(path string) error {
 	return errors.New("rtmr subsystem does not support RemoveAll")
 }
 
-func readTdx(entry *rtmrEntry, attr string) ([]byte, error) {
-	if !entry.initialized {
-		return nil, os.ErrNotExist
-	}
-	switch attr {
-	case tsmRtmrDigest:
-		return entry.rtmrMaps[entry.rtmrIndex].digest, nil
-	case tsmPathIndex:
-		return []byte(strconv.Itoa(entry.rtmrIndex)), nil
-	case tsmPathTcgMap:
-		return entry.rtmrMaps[entry.rtmrIndex].tcgMap, nil
-	}
-	return nil, os.ErrNotExist
+func readTdx(entry string, attr string) ([]byte, error) {
+	return os.ReadFile(path.Join(entry, attr))
 }
 
-func writeTdx(entry *rtmrEntry, attr string, content []byte) error {
+func writeTdx(entry string, attr string, content []byte) error {
 	switch attr {
 	case tsmRtmrDigest:
+		// Check if the content is a valid SHA384 hash.
 		if len(content) != crypto.SHA384.Size() {
 			return syscall.EINVAL
 		}
-		if !entry.initialized {
-			return os.ErrNotExist
+		// Check if the entry is initialized.
+		content, err := os.ReadFile(filepath.Join(entry, tsmPathIndex))
+		if err != nil {
+			return err
 		}
-		// According to the TDX module spec, userspace can only extend rtmr2 or rtmr3
-		if entry.rtmrIndex != 2 && entry.rtmrIndex != 3 {
+		rtmrIndex, err := strconv.Atoi(string(content))
+		if err != nil {
+			return err
+		}
+		if rtmrIndex != 2 && rtmrIndex != 3 {
 			return os.ErrPermission
 		}
-		oldDigest := entry.rtmrMaps[entry.rtmrIndex].digest
+		oldDigest, err := os.ReadFile(filepath.Join(entry, tsmRtmrDigest))
+		if err != nil {
+			return err
+		}
 		newDigest := sha512.Sum384(append(oldDigest[:], content...))
-		entry.rtmrMaps[entry.rtmrIndex].digest = newDigest[:]
+		if err := os.WriteFile(filepath.Join(entry, tsmRtmrDigest), newDigest[:], 0666); err != nil {
+			return err
+		}
 	case tsmPathIndex:
-		index, e := strconv.Atoi(string(content))
+		rtmrIndex, e := strconv.Atoi(string(content))
 		if e != nil {
 			return fmt.Errorf("WriteTdx: %v", e)
 		}
-		entry.rtmrIndex = index
-		entry.initialized = true
```

Comment:
You missed behavior before about only allowing initialization once. Please fix that. If the index has been set, then return `syscall.EBUSY`.

Reply:
Done. Note that behavior was not tested in rtmr_test.go. A proper implementation of rtmr.go should never allow such situations happen. I will add fakertmr_test.go in the future to test the behavior.